### ISSUE TITLE
AttnProcessor2_0 Significantly Slower

### ISSUE THREAD

**Issue by fpgaminer** [huggingface/diffusers#3493]
### Describe the bug

The new AttnProcessor2_0 makes my finetuning code run at 3.8 it/s, whereas if I force AttnProcessor the speed is 5.8 it/s. This is without `torch.compile`.  With `torch.compile` AttnProcessor2_0 is even worse at 3.5 it/s, but AttnProcessor remains at about 5.8 it/s.

Since training code can be quite complex I decided to write the minimal reproduction code as inference code. The difference is not as large for inference on my system. See the logs below. I'll try to write a minimal training reproduction soon.

### Reproduction

```
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
import torch
from diffusers import DiffusionPipeline
from diffusers.models.attention_processor import AttnProcessor, AttnProcessor2_0
import time

pipe = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.unet.set_attn_processor(AttnProcessor())

times = []
images = []
generator = torch.Generator(device="cuda").manual_seed(42)
prompt = "a photo of an astronaut riding a horse on mars"

for _ in range(16):
	start_time = time.time()
	image = pipe(prompt, generator=generator).images[0]
	times.append(time.time() - start_time)
	images.append(image)

print("Average time:", sum(times) / len(times))
print("Average time excluding first:", sum(times[1:]) / (len(times) - 1))

pipe.unet.set_attn_processor(AttnProcessor2_0())
times = []
images = []
generator = torch.Generator(device="cuda").manual_seed(42)
for _ in range(16):
	start_time = time.time()
	image = pipe(prompt, generator=generator).images[0]
	times.append(time.time() - start_time)
	images.append(image)

print("Average time:", sum(times) / len(times))
print("Average time excluding first:", sum(times[1:]) / (len(times) - 1))
```

### Logs

AttnProcessor2_0:
```shell
Average time: 6.953957229852676
Average time excluding first: 6.906311289469401
```

AttnProcessor:
```shell
Average time: 6.470208257436752
Average time excluding first: 6.470679505666097
```

### System Info

Dockerfile used for environment:

```
FROM pytorch/pytorch:2.0.1-cuda11.7-cudnn8-devel

ENV DEBIAN_FRONTEND=noninteractive PIP_PREFER_BINARY=1 PIP_NO_CACHE_DIR=1

RUN apt-get update && \
	apt-get install -y wget git && \
	apt-get clean && \
	rm -rf /var/lib/apt/lists/*

RUN pip install -U \
	requests==2.30.0 \
	Pillow==9.5.0 \
	transformers==4.29.2 \
	diffusers==0.16.1 \
	wandb==0.15.3 \
	datasets==2.12.0 \
	bitsandbytes==0.38.1 \
	accelerate==0.19.0

COPY Train.py /

ENTRYPOINT ["python", "/Train.py"]
```

**nvidia-smi**
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.105.17   Driver Version: 525.105.17   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:05:00.0 Off |                  N/A |
|  0%   31C    P8    27W / 420W |      5MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  NVIDIA GeForce ...  Off  | 00000000:0B:00.0  On |                  N/A |
|  0%   52C    P2   399W / 420W |  21885MiB / 24576MiB |     99%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

**Comment by fpgaminer** [comment]
Here's a minimal reproduction for the training loop:

```
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "1"
import torch
import torch.nn.functional as F
from diffusers import DiffusionPipeline
from diffusers.models.attention_processor import AttnProcessor, AttnProcessor2_0
import time
from tqdm import tqdm


total_steps = 256
batch_size = 1

pipe = DiffusionPipeline.from_pretrained("CompVis/stable-diffusion-v1-4")
pipe = pipe.to("cuda")
pipe.text_encoder.requires_grad_(False)
pipe.unet.requires_grad_(True)
pipe.vae.requires_grad_(False)
pipe.unet.set_attn_processor(AttnProcessor2_0())

input_ids = [pipe.tokenizer(["A photo of an astronaut riding a goat"], truncation=True, return_length=True, return_overflowing_tokens=False, padding=False, add_special_tokens=False, max_length=77).input_ids for x in range(batch_size)]

scaler = torch.cuda.amp.GradScaler(enabled=True)

start_time = time.time()
for device_step in tqdm(range(total_steps)):
	pipe.unet.train()

	latents = torch.randn((batch_size, 4, 64, 64), device='cuda', dtype=torch.float16)

	# Noise
	noise = torch.randn_like(latents)
	timesteps = torch.randint(0, pipe.scheduler.config.num_train_timesteps, (batch_size,), device=latents.device)
	timesteps = timesteps.long()

	noisy_latents = pipe.scheduler.add_noise(latents, noise, timesteps)

	# Conditioning
	encoder_hidden_states = [pipe.text_encoder(torch.asarray(input_id).to('cuda'), output_hidden_states=True).last_hidden_state[0] for input_id in input_ids]
	encoder_hidden_states = torch.stack(tuple(encoder_hidden_states))

	# Forward pass
	with torch.autocast('cuda', enabled=True):
		model_pred = pipe.unet(noisy_latents, timesteps, encoder_hidden_states).sample

		# Loss
		loss = F.mse_loss(model_pred.float(), noise.float(), reduction="mean")
		loss = loss / total_steps

	# Backward pass
	scaler.scale(loss).backward()

end_time = time.time()
print()
print("Done")
print("Time:", end_time - start_time)
print("Average time per step:", (end_time - start_time) / total_steps)
```

I get the following results on my setup:

**AttnProcessor2_0**
```
Time: 64.04590606689453
Average time per step: 0.25017932057380676
```

**AttnProcessor**
```
Time: 42.04754042625427
Average time per step: 0.16424820479005575
```

**Comment by williamberman** [comment]
tl;dr perf issues like these are complicated and might need a little elbow grease :grin:


hey @fpgaminer I'm having trouble reproducing this. When I use the example inference script, I get the 2.0 attention processor is faster.

```
AttentionProcessor
Average time: 7.894478380680084
Average time excluding first: 7.553478479385376

AttentionProcessor2.0
Average time: 6.9629515409469604
Average time excluding first: 6.969301064809163
```

What GPU model are you using? I don't see an obvious output in the nvidia-smi dump. It could very well be whatever model you have's compute capability has some issue with the optimized memory efficient attention kernels (though I would be surprised here as you would still get memory write savings and those compute capabilities would maybe impact the non-memory optimized attention processor?)

It might be hard for us to debug for you here, however I can give you some pointers. Instead of taking measurements over the entire pipeline, measure just the attention operations at the different resolutions in the unet and use the torch built in time measurement tools as well.

I have an example script here that I used for some benchmarks of the different memory efficient attention kernels, maybe you could adapt and re-purpose it here as well?

```py
import torch
from torch import nn
from torch.utils.benchmark import Timer, Compare
from torch.backends.cuda import sdp_kernel

torch.manual_seed(0)
torch.set_float32_matmul_precision("high")

num_heads = 8
device = "cuda"
num_threads = torch.get_num_threads()
min_run_time = 1

seq_len = 64 * 64
embed_dim_per_head = 40
dtype = torch.float16

embed_dim = embed_dim_per_head * num_heads

attn = nn.MultiheadAttention(
    embed_dim=embed_dim, num_heads=num_heads, batch_first=True, dtype=dtype, device=device
)

feature_map = torch.randn((1, seq_len, embed_dim), dtype=dtype, device=device)

def benchmark_fn():
    attn(feature_map, feature_map, feature_map, need_weights=False)

args = dict(stmt="benchmark_fn()", setup="from __main__ import benchmark_fn",
            num_threads=num_threads, sub_label=str(dtype))

with sdp_kernel(enable_flash=True, enable_mem_efficient=False, enable_math=False):
    flash_out = Timer(**args, description="flash").blocked_autorange(min_run_time=min_run_time)

with sdp_kernel(enable_flash=False, enable_mem_efficient=True, enable_math=False):
    mem_out = Timer(**args, description="mem").blocked_autorange(min_run_time=min_run_time)

with sdp_kernel(enable_mem_efficient=False, enable_flash=False, enable_math=True):
    math_out = Timer(**args, description="math").blocked_autorange(min_run_time=min_run_time)

Compare([flash_out, mem_out, math_out]).print()
```

**Comment by fpgaminer** [comment]
Oops, sorry, I didn't double check the `nvidia-smi` output.  I'm running on 3090.  And what's odd is that the SDPA kernels are faster in other applications, like ViT, both for inference and training.  I'll give your benchmark script a try when I get a chance.

**Comment by sayakpaul** [comment]
+1 to @williamberman's comment.

FWIW, I think SDPA speedup also depends a lot on the sequence of the operations and their respective kernels are dispatched. Sometimes, the speedup is not a free lunch and we might have to make some tweaks to make the speedup come to fruition. 

https://pytorch.org/tutorials/intermediate/scaled_dot_product_attention_tutorial.html#using-sdpa-with-torch-compile gives a good rundown of this.

**Comment by sunhs** [comment]
I also observed this situation, in the case of training, where `mem_efficient_sdp` got slow in the backward pass.

I'm using RTX 2080Ti, with pytorch 2.0.1, diffusers 0.17.0.

Here's the code to train a single `Attention` module.

```python
import argparse

import torch
import torch.nn.functional as F
from torch.backends.cuda import (
    enable_flash_sdp,
    enable_math_sdp,
    enable_mem_efficient_sdp,
    sdp_kernel,
)

from diffusers.models.attention_processor import (
    Attention,
    AttnProcessor,
    AttnProcessor2_0,
)

parser = argparse.ArgumentParser(formatter_class=argparse.ArgumentDefaultsHelpFormatter)
parser.add_argument("--a2", action="store_true")
parser.add_argument(
    "--sdpa-kernel", type=str, choices=["flash", "effi", "math"], default=None
)
args = parser.parse_args()


attn = Attention(
    320,
    heads=1,
    dim_head=320,
    bias=True,
    upcast_softmax=True,
    norm_num_groups=32,
    residual_connection=True,
    _from_deprecated_attn_block=True,
)
attn = attn.half().cuda()
attn.train()

if args.a2:
    attn.set_processor(AttnProcessor2_0())
    if args.sdpa_kernel is not None:
        enable_flash_sdp(False)
        enable_mem_efficient_sdp(False)
        enable_math_sdp(False)
    if args.sdpa_kernel == "flash":
        enable_flash_sdp(True)
    if args.sdpa_kernel == "effi":
        enable_mem_efficient_sdp(True)
    if args.sdpa_kernel == "math":
        enable_math_sdp(True)
else:
    attn.set_processor(AttnProcessor())

scaler = torch.cuda.amp.GradScaler(enabled=True)


forward_elapsed = 0
loss_elapsed = 0
backward_elapsed = 0

forward_elapsed_except_first_run = 0
loss_elapsed_except_first_run = 0
backward_elapsed_except_first_run = 0

for i in range(500):
    data = torch.randn(1, 4096, 320).half().cuda()
    with torch.autocast("cuda", torch.half):
        forward_start = torch.cuda.Event(enable_timing=True)
        forward_end = torch.cuda.Event(enable_timing=True)
        forward_start.record()
        output = attn(data)
        forward_end.record()
        torch.cuda.synchronize()
        forward_elapsed += forward_start.elapsed_time(forward_end)
        if i > 0:
            forward_elapsed_except_first_run += forward_start.elapsed_time(forward_end)

        loss_start = torch.cuda.Event(enable_timing=True)
        loss_end = torch.cuda.Event(enable_timing=True)
        loss_start.record()
        loss = F.mse_loss(output.float(), data.float(), reduction="mean")
        loss_end.record()
        torch.cuda.synchronize()
        loss_elapsed += loss_start.elapsed_time(loss_end)
        if i > 0:
            loss_elapsed_except_first_run += loss_start.elapsed_time(loss_end)

    backward_start = torch.cuda.Event(enable_timing=True)
    backward_end = torch.cuda.Event(enable_timing=True)
    backward_start.record()
    scaler.scale(loss).backward()
    backward_end.record()
    torch.cuda.synchronize()
    backward_elapsed += backward_start.elapsed_time(backward_end)
    if i > 0:
        backward_elapsed_except_first_run += backward_start.elapsed_time(backward_end)


print(
    f"forward: {forward_elapsed/1000:.3f}s, loss: {loss_elapsed/1000:.3f}s, backward: {backward_elapsed/1000:.3f}s"
)
print(
    f"Except first Run. forward: {forward_elapsed_except_first_run/1000:.3f}s, loss: {loss_elapsed_except_first_run/1000:.3f}s, backward: {backward_elapsed_except_first_run/1000:.3f}s"
)

```

Here's the result:

```
# Using AttnProcessor
python test.py
forward: 1.815s, loss: 0.089s, backward: 1.805s
Except first Run. forward: 1.005s, loss: 0.088s, backward: 1.799s

# Using AttnProcessor2_0 and mem_efficient_sdp
python test.py --a2 --sdpa-kernel effi
forward: 1.741s, loss: 0.094s, backward: 68.035s
Except first Run. forward: 0.931s, loss: 0.093s, backward: 67.843s

# Using AttnProcessor2_0 and math_sdp
python test.py --a2 --sdpa-kernel math
forward: 1.511s, loss: 0.090s, backward: 1.406s
Except first Run. forward: 0.741s, loss: 0.089s, backward: 1.401s

# flash_sdp is not supported in my case, so ignore it
```

@fpgaminer Could you confirm this on your machine ? If so, I think we could create an issue in pytorch.
